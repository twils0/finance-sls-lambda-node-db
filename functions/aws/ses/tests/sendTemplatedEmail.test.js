const AWS = require('aws-sdk'); // eslint-disable-line

const sesConfig = require('../ses.config.json');

const sendEmailExport = require('../sendTemplatedEmail');

const sendEmailModule = sendEmailExport.sendTemplatedEmail;

const sendTemplatedEmail = jest.fn();

global.console.log = jest.fn();

AWS.SES = jest.fn(() => ({
  sendTemplatedEmail,
}));

const from = `${sesConfig.fromName} <${sesConfig.fromVerifiedEmail}>`;
const replyTo = `${sesConfig.replyToName} <${sesConfig.replyToVerifiedEmail}>`;

const template = 'verificationEmailTemplate';
const emailAddresses = ['toTest0@test.com', 'toTest1@test.com'];
const data = { test: 'testData' };

const expectedParamsData0 = {
  Template: template,
  Source: from,
  ReplyToAddresses: [replyTo],
  TemplateData: JSON.stringify(data),
  Destination: { ToAddresses: [emailAddresses[0]] },
};
const expectedParamsData1 = {
  Template: template,
  Source: from,
  ReplyToAddresses: [replyTo],
  TemplateData: JSON.stringify(data),
  Destination: { ToAddresses: [emailAddresses[1]] },
};
const expectedParamsNoData0 = {
  Template: template,
  Source: from,
  ReplyToAddresses: [replyTo],
  TemplateData: '',
  Destination: { ToAddresses: [emailAddresses[0]] },
};
const expectedParamsNoData1 = {
  Template: template,
  Source: from,
  ReplyToAddresses: [replyTo],
  TemplateData: '',
  Destination: { ToAddresses: [emailAddresses[1]] },
};

describe('functions', () => {
  describe('aws', () => {
    describe('ses', () => {
      describe('sendTemplatedEmail', () => {
        afterEach(() => {
          global.console.log.mockReset();
          sendTemplatedEmail.mockReset();
        });

        it('fails and console logs email not sent, no email addresses, error', async () => {
          const result = await sendEmailModule(template, data, []);

          expect(result).toEqual(null);
          await setTimeout(() => {}, 10); // work-around to test that logout is or is not fired
          expect(sendTemplatedEmail).not.toBeCalled();
          expect(global.console.log).toMatchSnapshot();
        });

        it('fails and console logs failed to send error, when sendTemplatedEmail throws error', async () => {
          const error = {
            code: 'testCode',
            message: 'testMessage',
          };
          sendTemplatedEmail.mockImplementationOnce((params, callback) => {
            expect(params).toEqual(expectedParamsData0);
            callback(error, null);
          });

          const result = await sendEmailModule(template, data, emailAddresses);

          expect(result).toEqual(null);
          await setTimeout(() => {}, 10); // work-around to test that logout is or is not fired
          expect(sendTemplatedEmail).toHaveBeenCalledTimes(2);
          expect(global.console.log).toMatchSnapshot();
        });

        it('correct sends emails when provided data', async () => {
          sendTemplatedEmail.mockImplementationOnce((params, callback) => {
            expect(params).toEqual(expectedParamsData0);
            callback(null, null);
          });
          sendTemplatedEmail.mockImplementationOnce((params, callback) => {
            expect(params).toEqual(expectedParamsData1);
            callback(null, null);
          });

          const result = await sendEmailModule(template, data, emailAddresses);

          expect(result).toEqual(null);
          await setTimeout(() => {}, 10); // work-around to test that logout is or is not fired
          expect(sendTemplatedEmail).toHaveBeenCalledTimes(2);
          expect(global.console.log).toMatchSnapshot();
        });

        it('correct sends emails when provided data', async () => {
          sendTemplatedEmail.mockImplementationOnce((params, callback) => {
            expect(params).toEqual(expectedParamsNoData0);
            callback(null, null);
          });
          sendTemplatedEmail.mockImplementationOnce((params, callback) => {
            expect(params).toEqual(expectedParamsNoData1);
            callback(null, null);
          });

          const result = await sendEmailModule(template, null, emailAddresses);

          expect(result).toEqual(null);
          expect(sendTemplatedEmail).toHaveBeenCalledTimes(2);
          expect(global.console.log).toMatchSnapshot();
        });
      });
    });
  });
});
