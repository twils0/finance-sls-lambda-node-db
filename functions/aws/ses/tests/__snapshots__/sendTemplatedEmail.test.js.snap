// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functions aws ses sendTemplatedEmail correct sends emails when provided data 1`] = `[MockFunction]`;

exports[`functions aws ses sendTemplatedEmail correct sends emails when provided data 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "unexpected error - failed to send email - toTest0@test.com",
      "
",
      [Error: [2mexpect([22m[31mreceived[39m[2m).toEqual([22m[32mexpected[39m[2m)[22m

Expected value to equal:
  [32m{"Destination": {"ToAddresses": ["toTest0@test.com"]}, "ReplyToAddresses": ["Example <support@example.com>"], "Source": "Example <support@example.com>", "Template": "verificationEmailTemplate", "TemplateData": ""}[39m
Received:
  [31m{"Destination": {"ToAddresses": ["toTest0@test.com"]}, "ReplyToAddresses": ["Example <support@example.com>"], "Source": "Example <support@example.com>", "Template": "verificationEmailTemplate", "TemplateData": "{}"}[39m

Difference:

[32m- Expected[39m
[31m+ Received[39m

[33m@@ -7,7 +7,7 @@[39m
[2m    "ReplyToAddresses": Array [[22m
[2m      "Example <support@example.com>",[22m
[2m    ],[22m
[2m    "Source": "Example <support@example.com>",[22m
[2m    "Template": "verificationEmailTemplate",[22m
[32m-   "TemplateData": "",[39m
[31m+   "TemplateData": "{}",[39m
[2m  }[22m],
    ],
    Array [
      "unexpected error - failed to send email - toTest1@test.com",
      "
",
      [Error: [2mexpect([22m[31mreceived[39m[2m).toEqual([22m[32mexpected[39m[2m)[22m

Expected value to equal:
  [32m{"Destination": {"ToAddresses": ["toTest1@test.com"]}, "ReplyToAddresses": ["Example <support@example.com>"], "Source": "Example <support@example.com>", "Template": "verificationEmailTemplate", "TemplateData": ""}[39m
Received:
  [31m{"Destination": {"ToAddresses": ["toTest1@test.com"]}, "ReplyToAddresses": ["Example <support@example.com>"], "Source": "Example <support@example.com>", "Template": "verificationEmailTemplate", "TemplateData": "{}"}[39m

Difference:

[32m- Expected[39m
[31m+ Received[39m

[33m@@ -7,7 +7,7 @@[39m
[2m    "ReplyToAddresses": Array [[22m
[2m      "Example <support@example.com>",[22m
[2m    ],[22m
[2m    "Source": "Example <support@example.com>",[22m
[2m    "Template": "verificationEmailTemplate",[22m
[32m-   "TemplateData": "",[39m
[31m+   "TemplateData": "{}",[39m
[2m  }[22m],
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`functions aws ses sendTemplatedEmail fails and console logs email not sent, no email addresses, error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "email not sent - no email addresses provided",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`functions aws ses sendTemplatedEmail fails and console logs failed to send error, when sendTemplatedEmail throws error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "unexpected error - failed to send email - toTest0@test.com",
      "
",
      Object {
        "code": "testCode",
        "message": "testMessage",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
